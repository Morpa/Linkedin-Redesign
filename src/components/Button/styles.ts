import styled, { css, DefaultTheme } from 'styled-components'

import { ButtonProps } from '.'

export type WrapperProps = Pick<
  ButtonProps,
  'size' | 'color' | 'border' | 'iconLeft'
>

const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    width: 3.6rem;
    height: 3.6rem;
    border-radius: ${theme.border.radius};
    background: ${theme.colors.white};

    > svg {
      height: 2.4rem;
      width: 2.4rem;
      margin: 0;
    }
  `,
  medium: (theme: DefaultTheme) => css`
    width: 13.2rem;
    height: 3.6rem;
    border-radius: ${theme.border.radius};
  `,
  large: (theme: DefaultTheme) => css`
    width: 17rem;
    height: 3.2rem;
    border-radius: ${theme.border.radius};
  `,
  hasBorder: (theme: DefaultTheme) => css`
    border: 1px solid ${theme.colors.blue};
    background: ${theme.colors.white};
    width: 17rem;
    height: 3.2rem;
    border-radius: ${theme.border.radius};

    > small {
      color: ${theme.colors.blue};
    }
  `,

  white: (theme: DefaultTheme) => css`
    background: ${theme.colors.white};

    > small {
      color: ${theme.colors.black};
    }
  `,

  blue: (theme: DefaultTheme) => css`
    background: ${theme.colors.blue};

    > svg {
      color: ${theme.colors.white};
    }

    > small {
      color: ${theme.colors.white};
    }
  `
}

export const Wrapper = styled.button<WrapperProps>`
  ${({ theme, color, size, border, iconLeft }) => css`
    border: none;
    padding: ${theme.spacings.xxxsmall};
    outline: none;
    line-height: ${theme.spacings.small};
    cursor: pointer;
    transition: 0.1s;

    display: flex;
    flex-direction: ${iconLeft ? 'row-reverse' : ''};
    align-items: center;
    justify-content: space-between;

    > small {
      font-size: ${theme.font.sizes.small};
      flex: 1;
      text-transform: uppercase;
      font-weight: ${theme.font.normal};
    }

    &:hover {
      transform: scale(1.03);
    }

    &:active {
      transform: scale(0.98);
    }

    > svg {
      height: 1.6rem;
      width: 1.6rem;
      margin-right: ${theme.spacings.xxsmall};
      margin-left: ${iconLeft ? '0.8rem' : ''};
    }
    ${border && wrapperModifiers.hasBorder(theme)}
    ${!!size && wrapperModifiers[size](theme)}
    ${!!color && wrapperModifiers[color](theme)}
  `}
`
