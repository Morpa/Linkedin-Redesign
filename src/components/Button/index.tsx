import { AnchorHTMLAttributes, ButtonHTMLAttributes, forwardRef } from 'react'

import * as S from './styles'
import { WrapperProps } from './styles'

type ButtonTypes =
  | ButtonHTMLAttributes<HTMLButtonElement>
  | AnchorHTMLAttributes<HTMLAnchorElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  label?: string
  color?: 'white' | 'blue'
  border?: boolean
  iconLeft?: boolean
  icon?: React.ReactNode
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<WrapperProps, ButtonProps> = (
  {
    size = 'medium',
    icon,
    color = 'blue',
    border,
    children,
    iconLeft,
    ...props
  },
  ref
) => {
  return (
    <S.Wrapper
      size={size}
      color={color}
      border={border}
      iconLeft={iconLeft}
      ref={ref}
      {...props}
    >
      {!!icon && icon}
      {!!children && <small>{children}</small>}
    </S.Wrapper>
  )
}

export default forwardRef(Button)
