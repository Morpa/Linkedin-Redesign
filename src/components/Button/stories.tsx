import { Story, Meta } from '@storybook/react/types-6-0'
import { Edit, MoreVert } from '@styled-icons/material-outlined'

import Button, { ButtonProps } from '.'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      type: ''
    }
  }
} as Meta

export const basicBlue: Story<ButtonProps> = (args) => <Button {...args} />
basicBlue.args = {
  children: 'contact info',
  color: 'blue',
  size: 'large'
}

export const basicWhite: Story<ButtonProps> = (args) => <Button {...args} />
basicWhite.args = {
  children: 'edit profile',
  color: 'white',
  icon: <Edit />
}

export const withBorder: Story<ButtonProps> = (args) => <Button {...args} />
withBorder.args = {
  children: '1.090 connections',
  border: true
}

export const small: Story<ButtonProps> = (args) => <Button {...args} />
small.args = {
  icon: <MoreVert />,
  size: 'small'
}

export const iconLeft: Story<ButtonProps> = (args) => (
  <Button iconLeft {...args} />
)
iconLeft.args = {
  icon: <MoreVert />,
  children: 'Questions?',
  size: 'large',
  color: 'blue'
}
