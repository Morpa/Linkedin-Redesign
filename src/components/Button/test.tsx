import { screen } from '@testing-library/react'
import { Edit } from '@styled-icons/material-outlined'

import { renderWithTheme } from 'utils/tests/helpers'

import Button from '.'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(
      <Button color="blue">My Button</Button>
    )

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      height: '3.6rem',
      width: '13.2rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the small size', () => {
    renderWithTheme(<Button size="small">My Button</Button>)

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      height: '3.6rem',
      width: '3.6rem'
    })
  })

  it('should render the large size', () => {
    renderWithTheme(<Button size="large">My Button</Button>)

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      height: '3.2rem',
      width: '17rem'
    })
  })

  it('should render the button with icon', () => {
    renderWithTheme(
      <Button icon={<Edit data-testid="icon" />}>My Button</Button>
    )

    expect(screen.getByText(/my button/i)).toBeInTheDocument()

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render the button with white color', () => {
    renderWithTheme(<Button color="white">My Button</Button>)

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      background: '#FFFFFF'
    })
  })

  it('should render the button with border', () => {
    renderWithTheme(
      <Button border color="white">
        My Button
      </Button>
    )

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      border: '1px solid #0275B1',
      background: '#FFFFFF'
    })
  })

  it('should render Button as a link', () => {
    renderWithTheme(
      <Button as="a" href="/link">
        Buy now
      </Button>
    )

    expect(screen.getByRole('link', { name: /buy now/i })).toHaveAttribute(
      'href',
      '/link'
    )
  })

  it('should render the button with icon reverse', () => {
    renderWithTheme(
      <Button icon={<Edit data-testid="icon" />} iconLeft>
        My Button
      </Button>
    )

    expect(screen.getByRole('button', { name: /My Button/i })).toHaveStyle({
      'flex-direction': 'row-reverse'
    })

    expect(screen.getByText(/my button/i)).toBeInTheDocument()

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
})
