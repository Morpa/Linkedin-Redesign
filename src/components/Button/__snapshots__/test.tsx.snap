// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size by default 1`] = `
.c0 {
  border: none;
  padding: 0.4rem;
  outline: none;
  line-height: 2.4rem;
  cursor: pointer;
  -webkit-transition: 0.1s;
  transition: 0.1s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 13.2rem;
  height: 3.6rem;
  border-radius: 0.4rem;
  background: #0275B1;
}

.c0 > small {
  font-size: 1.4rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  text-transform: uppercase;
  font-weight: 400;
}

.c0:hover {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c0:active {
  -webkit-transform: scale(0.98);
  -ms-transform: scale(0.98);
  transform: scale(0.98);
}

.c0 > svg {
  height: 1.6rem;
  width: 1.6rem;
  margin-right: 0.8rem;
}

.c0 > svg {
  color: #FFFFFF;
}

.c0 > small {
  color: #FFFFFF;
}

<button
  class="c0"
  color="blue"
>
  <small>
    My Button
  </small>
</button>
`
